include/albatross/src/models/patchwork_gp.hpp

```diff
@@ -0,0 +1,555 @@
+
+  std::vector<GroupFeatureType> group_features;
+
+  auto emplace_in_output = [&](const auto &key, const auto &features) {
+    for (const auto &f : features) {
+      group_features.emplace_back(as_group_feature(key, f));
+    }
+  };
+  grouped_features.apply(emplace_in_output);
+
+  return group_features;
+}
+
+template <typename GroupKey, typename FeatureType>
+auto as_group_features(const GroupKey &key,
+                       const std::vector<FeatureType> &features) {
+  using GroupFeatureType =
+      GroupFeature<GroupKey, typename std::decay<FeatureType>::type>;
+
+  std::vector<GroupFeatureType> group_features;
+  for (const auto &f : features) {
+    group_features.emplace_back(as_group_feature(key, f));
+  }
+  return group_features;
+}
+
+template <typename SubCaller> struct PatchworkCallerBase {
+  template <
+      typename CovFunc, typename X, typename Y,
+      typename std::enable_if<
+          has_valid_cov_caller<CovFunc, SubCaller, X, Y>::value, int>::type = 0>
+  static double call(const CovFunc &cov_func, const X &x, const Y &y) {
+    return SubCaller::call(cov_func, x, y);
+  }
+
+  template <typename CovFunc, typename GroupKey, typename FeatureTypeX,
+            typename FeatureTypeY>
+  static double call(const CovFunc &cov_func,
+                     const GroupFeature<GroupKey, FeatureTypeX> &x,
+                     const GroupFeature<GroupKey, FeatureTypeY> &y) {
+    if (x.key == y.key) {
+      return SubCaller::call(cov_func, x.feature, y.feature);
+    } else {
+      return 0.;
+    }
+  }
+
+  template <typename CovFunc, typename GroupKey, typename FeatureTypeX,
+            typename FeatureTypeY>
+  static double call(const CovFunc &cov_func,
+                     const GroupFeature<GroupKey, FeatureTypeX> &x,
+                     const BoundaryFeature<GroupKey, FeatureTypeY> &y) {
+    if (x.key == y.lhs) {
+      return SubCaller::call(cov_func, x.feature, y.feature);
+    } else if (x.key == y.rhs) {
+      return -SubCaller::call(cov_func, x.feature, y.feature);
+    } else {
+      return 0.;
+    }
+  }
+
+  template <typename CovFunc, typename GroupKey, typename FeatureTypeX,
+            typename FeatureTypeY>
+  static double call(const CovFunc &cov_func,
+                     const BoundaryFeature<GroupKey, FeatureTypeX> &x,
+                     const BoundaryFeature<GroupKey, FeatureTypeY> &y) {
+
+    if (x.lhs == y.lhs && x.rhs == y.rhs) {
+      return 2 * SubCaller::call(cov_func, x.feature, y.feature);
+    } else if (x.lhs == y.lhs && x.rhs != y.rhs) {
+      return SubCaller::call(cov_func, x.feature, y.feature);
+    } else if (x.lhs != y.lhs && x.rhs == y.rhs) {
+      return SubCaller::call(cov_func, x.feature, y.feature);
+    } else if (x.lhs == y.rhs && x.rhs != y.lhs) {
+      return -SubCaller::call(cov_func, x.feature, y.feature);
+    } else if (x.lhs != y.rhs && x.rhs == y.lhs) {
+      return -SubCaller::call(cov_func, x.feature, y.feature);
+    } else {
```

Comment:
Yeah good call, I guess the way boundary features are created this just never happens, but might as well cover it in case it somehow does in the future.

Reply:
Yeah you're correct, that should be defined, though if used as intended there should never be two `BoundaryFeatures` which are symmetric so we shouldn't hit that case.  Added it for completeness.